src/sortTemplate.ts

```diff
@@ -0,0 +1,297 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from "vscode";
+import { Language } from "../extension.bundle";
+import { templateKeys } from './constants';
+import { DeploymentTemplate } from "./DeploymentTemplate";
+import { ext } from './extensionVariables';
+import { IParameterDefinition } from './IParameterDefinition';
+import * as Json from "./JSON";
+import * as language from "./Language";
+import { UserFunctionDefinition } from './UserFunctionDefinition';
+import { UserFunctionNamespaceDefinition } from './UserFunctionNamespaceDefinition';
+import { IVariableDefinition } from './VariableDefinition';
+
+export enum SortType {
+    Resources,
+    Outputs,
+    Parameters,
+    Variables,
+    Functions,
+    TopLevel
+}
+
+export class SortQuickPickItem implements vscode.QuickPickItem {
+    public label: string;
+    public value: SortType;
+    public description: string;
+
+    constructor(label: string, value: SortType, description: string) {
+        this.label = label;
+        this.value = value;
+        this.description = description;
+    }
+}
+
+export function getQuickPickItems(): SortQuickPickItem[] {
+    let items: SortQuickPickItem[] = [];
+    items.push(new SortQuickPickItem("Functions", SortType.Functions, "Sort function namespaces and functions"));
+    items.push(new SortQuickPickItem("Outputs", SortType.Outputs, "Sort outputs"));
+    items.push(new SortQuickPickItem("Parameters", SortType.Parameters, "Sort parameters for the template"));
+    items.push(new SortQuickPickItem("Resources", SortType.Resources, "Sort resources based on the name including first level of child resources"));
+    items.push(new SortQuickPickItem("Variables", SortType.Variables, "Sort variables"));
+    items.push(new SortQuickPickItem("Top level", SortType.TopLevel, "Sort top level items based on recommended order (parameters, functions, variables, resources, outputs)"));
+    return items;
+}
+
+export async function sortTemplate(template: DeploymentTemplate | undefined, sortType: SortType): Promise<void> {
+    if (template === undefined) {
```

Comment:
Everywhere, please:

replace
if (x === undefined)
   with
if (!x)

and replace
if (x !== undefined)
with
if (x)

The reason is that the second form handles null the same as undefined.  This is generally a problem because we've mostly standardized on undefined, but is still safer practice.

Reply:
I replaced undefined everywhere except where it would generate a lint warning.

Reply:
Fixed

Reply:
Fixed